### DISCUSSION TITLE
How to debug coreclr with clion at the source level

### DISCUSSION THREAD

**Discussion by dacheng-gao** [dotnet/runtime#109097] [Q&A] [answered]
I have tried a lot and I can't set breakpoint inside coreclr, like `EEStartup` method.

> The answer is build with option `--keepnativesymbols true`.

What I have done so far:

### Get the code

Head to https://github.com/dotnet/runtime/releases , download .NET 9 RC 2 and unzip it.

### Full build in terminal

> Use clang@16, apple clang or brew clang both are working, but brew clang@19 will crash.

```
./build.sh --ninja
```

coreclr build artifacts are inside `artifacts/bin/coreclr/osx.arm64.Debug`, I'll call this dir `CORE_ROOT`.

~I need to copy libSystem.Native.dylib and System.Runtime.dll to the CORE_ROOT dir which mentioned above.~

### Open with CLion

> I'm using apple m3.

> Need to apply a patch of `CMakePresets.json` first, it's https://github.com/dotnet/runtime/pull/109061/files.

Open CLion and coreclr directory, it will recognize it as cmake project, and enable the preset profile `osx.arm64.Debug`.

#### Attempt 1 debug corerun`main

> Success.

Open target `corerun`, set program arguments to `/path/to/your/dotnet-app.dll`, and environment variable `CORE_ROOT=/path/to/artifacts/bin/coreclr/osx-arm64-Debug` and `CORE_LIBRARIES=/path/to/artifacts/bin/runtime/net9.0-osx-Debug-arm64`.

Now set the breakpoint for main function, hit Debug button, here we go:

![image](https://github.com/user-attachments/assets/4939b972-dd8a-43f6-8d69-46b6e71180fa)

#### Attempt 2 debug coreclr`EEStartup

> Failed.

Open target `coreclr`, set program arguments to `/path/to/your/dotnet-app.dll`, and environment variable `CORE_ROOT=/path/to/artifacts/bin/coreclr/osx-arm64-Debug` and `CORE_LIBRARIES=/path/to/artifacts/bin/runtime/net9.0-osx-Debug-arm64`.

Now set the breakpoint for EEStartup function, hit Debug button, here we go:

![image](https://github.com/user-attachments/assets/f32170db-42a5-4698-a1d8-d30b2023759a)

It just run to end, without hitting the breakpoint.

**Comment by huoyaoyuan** [comment]
The steps performed looks similar to [steps for Visual Studio](https://github.com/dotnet/runtime/blob/main/docs/workflow/debugging/coreclr/debugging-runtime.md#using-visual-studio). I'm not familiar with CLion.

> I need to copy libSystem.Native.dylib and System.Runtime.dll to the CORE_ROOT dir which mentioned above.

They are set by the environment variable CORE_LIBRARIES.

Some potential troubleshooting approaches:
- Does CLion has a "break point can't be set" message? When the symbols are missing/mismatched, or the dylib hasn't be loaded yet.
- What about single stepping through corerun?
- Does corerun runs everything successfully?

**Reply by dacheng-gao** [reply]
I skipped the part of 'steps for Visual Studio', now I'll give it a try.

By setting up `CORE_LIBRARIES` to `/path/to/artifacts/bin/runtime/net9.0-osx-Debug-arm64`, and delete all previously copied files, it worked well.

> Does CLion has a "break point can't be set" message? When the symbols are missing/mismatched, or the dylib hasn't be loaded yet.

No, I think symbols are all there, and dylib now loaded successfully, as it run well and step debugging working for `corerun`. But it indeed show the message below while running `corerun`:

![image](https://github.com/user-attachments/assets/c9b6a16b-1880-48a0-9be1-1b2c78e6a2d5)

> What about single stepping through corerun?

It's working fine.

> Does corerun runs everything successfully?

Yes, but can not step into like `coreclr_init_func`.

**Reply by huoyaoyuan** [reply]
> But it indeed show the message below while running `corerun`:

This should be the problem. CLion is not associating the symbol of libcoreclr.dylib with the source. It only correctly sets up debugging for the entry executable `corerun`.

It's now IDE specific stuff and you have to learn about how CLion loads debug symbol.

**Reply by dacheng-gao** [reply]
Thanks @huoyaoyuan for pointing out the new clue.

**Reply by dacheng-gao** [reply]
I got a work around here.

By adding / loading `.dwarf` inside lldb manually, it works.

![image](https://github.com/user-attachments/assets/4f817632-7ce8-44fb-8dca-7a1b97fa3999)

**Reply by dacheng-gao** [reply, marked as answer]
```
  [817/1739] Linking CXX shared library dlls/mscoree/coreclr/libcoreclr.dylib
  Stripping symbols from libcoreclr.dylib into libcoreclr.dylib.dwarf
```

I found these **Stripping** lines in the output of build.

I found there is an option to configure if keep the native symbols or not, which is `--keepnativesymbols`, but previously I can not build with this option, now I figured it out with option `--keepnativesymbols true`.

The complete build command is 

```
./build.sh --ninja --keepnativesymbols true
```

After build like this, breakpoint debugging inside coreclr works well.

**Reply by huoyaoyuan** [reply]
You can also try to explore how to make CLion load stripped symbol by default. It's often more useful to debug with the exact same build configuration used for the repository.

**Reply by dacheng-gao** [reply]
Sure.

I found this https://lldb.llvm.org/use/ondemand.html, I'll give it a try later. No lucky.

I found there is an issue open https://github.com/dotnet/runtime/issues/92911.

**Reply by EgorBo** [reply]
@successgo it'd be nice to have a documentation on how to setup CLion checked in. We don't have a good story on non-windows fior this
